File: path_finding.rb
#! /usr/bin/env ruby


# My reference to the algorithm:
# https://www.youtube.com/watch?v=KNXfSOx4eEE
# 
# and using the maze from maze.rb

class Graph
  attr_accessor :start_node, :goal_node, :nodes, :initialized

  def initialize
    @initialized = false
    @nodes = []
  end

  def initialized?
    @initialized
  end

  def add_node(node)
    @initialized = false

    node.graph = self

    if node.start_node?
      raise ArgumentError, 'already defined one' if @start_node
      @start_node = node
    end
    if node.goal_node?
      raise ArgumentError, 'already defined one' if @goal_node
      @goal_node = node
    end

    @nodes << node
  end

  def complete!
    @nodes.map(&:heuristic)

    @nodes.each do |node|
      # up
      up = @nodes.select{ |n| (n.x == node.x) && (n.y == node.y+1) }.first
      node.edges << Edge.new(up, 10) if up

      # up-left
      upl = @nodes.select{ |n| (n.x == node.x+1) && (n.y == node.y+1) }.first
      node.edges << Edge.new(upl, 14) if upl

      # up-right
      upr = @nodes.select{ |n| (n.x == node.x-1) && (n.y == node.y+1) }.first
      node.edges << Edge.new(upr, 14) if upr

      # down
      down = @nodes.select{ |n| (n.x == node.x) && (n.y == node.y-1) }.first
      node.edges << Edge.new(down, 10) if down

      # down-left
      downl = @nodes.select{ |n| (n.x == node.x-1) && (n.y == node.y-1) }.first
      node.edges << Edge.new(downl, 14) if downl

      # down-right
      downr = @nodes.select{ |n| (n.x == node.x+1) && (n.y == node.y-1) }.first
      node.edges << Edge.new(downr, 14) if downr

      # left
      left = @nodes.select{ |n| (n.x == node.x-1) && (n.y == node.y) }.first
      node.edges << Edge.new(left, 10) if left

      # right
      right = @nodes.select{ |n| (n.x == node.x+1) && (n.y == node.y) }.first
      node.edges << Edge.new(right, 10) if right
    end

    @initialized = true
  end

  def solve!

    self.start_node.cost = 0
    @open = [self.start_node]

    while(current_node = @open.sort!{|a,b| a.cost <=> b.cost }.shift) do

      puts "examining #{current_node.x}, #{current_node.y}"

      current_node.edges.each do |edge|

        puts "looking at edge going to #{edge.to.x}, #{edge.to.y} (cost : #{edge.movement}, heuristic: #{edge.to.heuristic}, class: #{edge.to.class})"

        # don't do anything
        if edge.to.visited? || edge.to.obstacle_node?

        elsif edge.to.goal_node?
          edge.to.parent = current_node
          puts "Goal Node Reach, path through: #{current_node.x}, #{current_node.y}"

        else
          total_cost = edge.to.heuristic + edge.movement + (current_node.cost)

          if @open.index(edge.to)
            # reassign
            if edge.to.cost > total_cost
              edge.to.cost = total_cost
              edge.to.parent = current_node
            end

          else
            # add to the open list, set an initial value from here, set a current parent
            @open << edge.to
            edge.to.cost = total_cost
            edge.to.parent = current_node
          end

        end
      end

      current_node.visited = true

    end
  end
end

class Node
  attr_accessor :visited, :parent, :edges, :x, :y, :cost, :graph

  def initialize(x,y)
    @x = x
    @y = y

    @visited = false
    @parent  = nil
    @edges = []
  end

  def start_node?
    false
  end

  def goal_node?
    false
  end

  def obstacle_node?
    false
  end

  def visited?
    @visited
  end

  def heuristic
    return @heuristic if @heuristic

    delta_x = (graph.goal_node.x - self.x).abs
    delta_y = (graph.goal_node.y - self.y).abs

    @heuristic = delta_x + delta_y
  end
end

class StartNode < Node
  def start_node?
    true
  end
end
class GoalNode < Node
  def goal_node?
    true
  end
  def heuristic
    0
  end
end
class ObstacleNode < Node
  def obstacle_node?
    true
  end
  def heuristic
    nil
  end
end

class Edge
  attr_accessor :to, :movement

  def initialize(to, movement)
    @to = to
    @movement = movement
  end
end

# MAZE LIKE THIS
# row, by row.
# astericks are walls/obstacles
# dashes are open space
# S is the start
# E is the goal

# '*','*','*','*','*','-','-','-'
graph = Graph.new

graph.add_node ObstacleNode.new(0,0)
graph.add_node ObstacleNode.new(1,0)
graph.add_node ObstacleNode.new(2,0)
graph.add_node ObstacleNode.new(3,0)
graph.add_node ObstacleNode.new(4,0)
graph.add_node Node.new(5,0)
graph.add_node Node.new(6,0)
graph.add_node Node.new(7,0)

# '*','-','-','-','*','-','-','-'
graph.add_node ObstacleNode.new(0,1)
graph.add_node Node.new(1,1)
graph.add_node Node.new(2,1)
graph.add_node Node.new(3,1)
graph.add_node ObstacleNode.new(4,1)
graph.add_node Node.new(5,1)
graph.add_node Node.new(6,1)
graph.add_node Node.new(7,1)

# '*','S','*','*','*','-','-','-'
graph.add_node ObstacleNode.new(0,2)
graph.add_node StartNode.new(1,2)
graph.add_node ObstacleNode.new(2,2)
graph.add_node ObstacleNode.new(3,2)
graph.add_node ObstacleNode.new(4,2)
graph.add_node Node.new(5,2)
graph.add_node Node.new(6,2)
graph.add_node Node.new(7,2)

# '*','-','-','-','*','*','*','*'
graph.add_node ObstacleNode.new(0,3)
graph.add_node Node.new(1,3)
graph.add_node Node.new(2,3)
graph.add_node Node.new(3,3)
graph.add_node ObstacleNode.new(4,3)
graph.add_node ObstacleNode.new(5,3)
graph.add_node ObstacleNode.new(6,3)
graph.add_node ObstacleNode.new(7,3)

# '*','-','*','-','-','-','-','*'
graph.add_node ObstacleNode.new(0,4)
graph.add_node Node.new(1,4)
graph.add_node ObstacleNode.new(2,4)
graph.add_node Node.new(3,4)
graph.add_node Node.new(4,4)
graph.add_node Node.new(5,4)
graph.add_node Node.new(6,4)
graph.add_node ObstacleNode.new(7,4)

# '*','-','-','-','*','-','-','*'
graph.add_node ObstacleNode.new(0,5)
graph.add_node Node.new(1,5)
graph.add_node Node.new(2,5)
graph.add_node Node.new(3,5)
graph.add_node ObstacleNode.new(4,5)
graph.add_node Node.new(5,5)
graph.add_node Node.new(6,5)
graph.add_node ObstacleNode.new(7,5)

# '*','*','*','*','*','-','E','*'
graph.add_node ObstacleNode.new(0,6)
graph.add_node ObstacleNode.new(1,6)
graph.add_node ObstacleNode.new(2,6)
graph.add_node ObstacleNode.new(3,6)
graph.add_node ObstacleNode.new(4,6)
graph.add_node Node.new(5,6)
graph.add_node GoalNode.new(6,6)
graph.add_node ObstacleNode.new(7,6)

# '-','-','-','-','*','*','*','*'
graph.add_node Node.new(0,7)
graph.add_node Node.new(1,7)
graph.add_node Node.new(2,7)
graph.add_node Node.new(3,7)
graph.add_node ObstacleNode.new(4,7)
graph.add_node ObstacleNode.new(5,7)
graph.add_node ObstacleNode.new(6,7)
graph.add_node ObstacleNode.new(7,7)

graph.complete!
graph.solve!

parent = graph.goal_node

while(parent = parent.parent) do
  puts "#{parent.class}: #{parent.x}, #{parent.y}"
end

File: path_finding.js
#!/usr/bin/env node --harmony

// recreating my ruby implementation in js

// first, i'm going to defined some classes to aid in representing a graph

// nodes, which may have a position and can be labelled
function Node(label, x, y) {
  this.label = label;
  this.edges = [];
  this.parent = null;
  this.heuristic = 0;
  this.cost = 0;
  this.x = x;
  this.y = y;
  this.visited = false;

  this.startNode = function() {
    return (this.label == 'start')
  }
  this.goalNode = function() {
    return (this.label == 'goal')
  }
  this.portalNode = function() {
    return (this.label == 'portal')
  }
  this.wallNode = function() {
    return (this.label == 'wall')
  }
}

// edges, a join which original from and point to a destination nodes
function Edge(destination, cost) {
  this.destination = destination;
  this.cost = cost;
}

// and the graph which we can ask questions of is made up of nodes and edges
function Graph() {
  this.complete = false;
  this.startNode = null;
  this.goalNode = null;
  this.solved = false;
  this.nodes = [];

  this.addNode = function(node) {
    this.complete = false;
    this.nodes.push(node);

    if (node.startNode()) {
      this.startNode = node;
    }
    if (node.goalNode()) {
      this.goalNode = node;
    }
  }

  this.initialize = function() {
    if (!this.startNode) {
      throw "I am adrift at sea without a starting node";
    }
    if (!this.goalNode) {
      throw "Without goal nodes, what is my purpose? Do I pass butter?";
    }

    var delta_x, delta_y;

    // calculate values on each node
    for(var i = 0, node; node = this.nodes[i]; i++) {
      // calculate the heuristic

      // for A*, afaik, our heuristic is the as the crow flies distance to the goal
      // which we can simply calculate as the sum of the absolute values of delta x and y
      delta_x = Math.abs(this.goalNode.x - node.x);
      delta_y = Math.abs(this.goalNode.y - node.y);

      node.heuristic = delta_x + delta_y;

      // add edges, in this case it is all the potential moves

      // this is obviously a naive and inefficient way to do this
      up    = this.nodes.filter(function(p) { return ((p.x == node.x ) && (p.y == node.y + 1 )) })[0];
      upl   = this.nodes.filter(function(p) { return ((p.x == node.x - 1) && (p.y == node.y + 1 )) })[0];
      upr   = this.nodes.filter(function(p) { return ((p.x == node.x + 1) && (p.y == node.y + 1)) })[0];
      down  = this.nodes.filter(function(p) { return ((p.x == node.x) && (p.y == node.y - 1)) })[0];
      downl = this.nodes.filter(function(p) { return ((p.x == node.x - 1) && (p.y == node.y - 1)) })[0];
      downr = this.nodes.filter(function(p) { return ((p.x == node.x + 1) && (p.y == node.y - 1)) })[0];
      left  = this.nodes.filter(function(p) { return ((p.x == node.x - 1) && (p.y == node.y )) })[0];
      right = this.nodes.filter(function(p) { return ((p.x == node.x + 1) && (p.y == node.y )) })[0];

      // left right cost 10
      if (up)    { node.edges.push(new Edge(up, 10)) }
      if (down)  { node.edges.push(new Edge(down, 10)) }
      if (left)  { node.edges.push(new Edge(left, 10)) }
      if (right) { node.edges.push(new Edge(right, 10)) }

      // diagonals cost sqrt(2) * 10
      if (upl)   { node.edges.push(new Edge(upl, 10 * Math.sqrt(2))) }
      if (upr)   { node.edges.push(new Edge(upr, 10 * Math.sqrt(2))) }
      if (downl) { node.edges.push(new Edge(downl, 10 * Math.sqrt(2))) }
      if (downr) { node.edges.push(new Edge(downr, 10 * Math.sqrt(2))) }

      // magic portals!
      if (node.portalNode()) {
        node.edges.push(new Edge(this.goalNode, 0))
      }

    }

    this.complete = true;
  }

  this.solve = function() {

    // the algorithm gets us to the goal node, and tries to do so efficiently
    // 
    // we get to the goal node, purely through the re-parenting procedure
    // which lets us trace our steps backward
    // 
    // to get us to pick our next moves intelligently, we consider two items:
    //    1. the movement costs, which are ~14.1 for diagonals and 10 points for ordinals
    //    2. the heuristic, the as-the-crow flies distance to the goal, which we cached
    // 
    // we arrive at the final cost by for each node by summing the previous movement costs
    // up to and including this point and the heuristic
    // 
    // the previous movement costs are the parents movement costs
    // in this simple case all of the movements are pretty standard, we're on a grid
    // but it's obvious in real life applications where money and time are factored into cost
    // or in a game where there could be a terrain cost (sand vs asphalt) it might
    // have been smarter to go around rather than through something
    // 
    // we account for that through re-parenting, every time a new node is explored we're
    // looking to see if other nodes we intend to visit would benefit from having stepped
    // first through this node. If this is the case, we re-parent and re-calculate that node's cost
    // 

    // LETS DO IT

    // first, the cost of the start node is just the heuristic
    this.startNode.cost = this.startNode.heuristic;

    // we have to keep track of what we've visited, and are considering visiting
    // we store the visited state on the node itself, so we can just ask the object
    // 
    // we start, of course, with the startNode
    var visiting = [this.startNode];
    var currentNode, relevantEdges, indexOfNode, totalCost;

    while(!this.solved) {
      // next, not relevant for the startNode
      // we'd like to examine these in order of lowest cost to greatest
      currentNode = visiting.sort(function(a,b) { return a.cost > b.cost }).shift();

      if (typeof currentNode == 'undefined') {
        throw 'ran out of nodes to check before we found a solution'
      }

      // exclude visited nodes and obstacles like walls
      relevantEdges = currentNode.edges.filter(function(e) { return !(e.destination.wallNode() || e.destination.visited) })

      // iterate over the edges
      for(var i=0, currentEdge; currentEdge = relevantEdges[i]; i++) {

        // so this is the meat and potatoes

        // first, is this in the visiting list?
        // remember JS quirk.. indexOf returns -1 if not in list
        indexOfNode = visiting.indexOf(currentEdge.destination);
        totalCost = currentNode.cost + currentEdge.destination.heuristic + currentEdge.cost;

        // it is not
        if (indexOfNode == -1) {

          // set the cost
          currentEdge.destination.parent = currentNode;
          currentEdge.destination.cost = totalCost;

          visiting.push(currentEdge.destination);

        // it is
        } else {

          if (currentEdge.destination.cost > totalCost) {
            // re-set the cost and parent
            currentEdge.destination.parent = currentNode;
            currentEdge.destination.cost = totalCost;
          }

        }

        if (currentEdge.destination.goalNode()) {
          this.solved = true
        }

      }

      // and now we've visited this node
      currentNode.visited = true;

    }

  }

}

// now we're going to instantiate a graph
var graph = new Graph();

// add nodes
graph.addNode(new Node('wall', 0, 0))
graph.addNode(new Node('wall', 1, 0))
graph.addNode(new Node('wall', 2, 0))
graph.addNode(new Node('wall', 3, 0))
graph.addNode(new Node('wall', 4, 0))
graph.addNode(new Node('corridor', 5, 0))
graph.addNode(new Node('corridor', 6, 0))
graph.addNode(new Node('corridor', 7, 0))

graph.addNode(new Node('wall', 0, 1))
graph.addNode(new Node('corridor', 1, 1))
graph.addNode(new Node('corridor', 2, 1))
graph.addNode(new Node('portal', 3, 1))
graph.addNode(new Node('wall', 4, 1))
graph.addNode(new Node('corridor', 5, 1))
graph.addNode(new Node('corridor', 6, 1))
graph.addNode(new Node('corridor', 7, 1))

graph.addNode(new Node('wall', 0, 2))
graph.addNode(new Node('start', 1, 2))
graph.addNode(new Node('wall', 2, 2))
graph.addNode(new Node('wall', 3, 2))
graph.addNode(new Node('wall', 4, 2))
graph.addNode(new Node('corridor', 5, 2))
graph.addNode(new Node('corridor', 6, 2))
graph.addNode(new Node('corridor', 7, 2))

graph.addNode(new Node('wall', 0, 3))
graph.addNode(new Node('corridor', 1, 3))
graph.addNode(new Node('corridor', 2, 3))
graph.addNode(new Node('corridor', 3, 3))
graph.addNode(new Node('wall', 4, 3))
graph.addNode(new Node('wall', 5, 3))
graph.addNode(new Node('wall', 6, 3))
graph.addNode(new Node('wall', 7, 3))

graph.addNode(new Node('wall', 0, 4))
graph.addNode(new Node('corridor', 1, 4))
graph.addNode(new Node('wall', 2, 4))
graph.addNode(new Node('corridor', 3, 4))
graph.addNode(new Node('corridor', 4, 4))
graph.addNode(new Node('corridor', 5, 4))
graph.addNode(new Node('corridor', 6, 4))
graph.addNode(new Node('wall', 7, 4))

graph.addNode(new Node('wall', 0, 5))
graph.addNode(new Node('corridor', 1, 5))
graph.addNode(new Node('corridor', 2, 5))
graph.addNode(new Node('corridor', 3, 5))
graph.addNode(new Node('wall', 4, 5))
graph.addNode(new Node('corridor', 5, 5))
graph.addNode(new Node('corridor', 6, 5))
graph.addNode(new Node('wall', 7, 5))

graph.addNode(new Node('wall', 0, 6))
graph.addNode(new Node('wall', 1, 6))
graph.addNode(new Node('wall', 2, 6))
graph.addNode(new Node('wall', 3, 6))
graph.addNode(new Node('wall', 4, 6))
graph.addNode(new Node('corridor', 5, 6))
graph.addNode(new Node('goal', 6, 6))
graph.addNode(new Node('wall', 7, 6))

graph.addNode(new Node('corridor', 0, 7))
graph.addNode(new Node('corridor', 1, 7))
graph.addNode(new Node('corridor', 2, 7))
graph.addNode(new Node('corridor', 3, 7))
graph.addNode(new Node('wall', 4, 7))
graph.addNode(new Node('wall', 5, 7))
graph.addNode(new Node('wall', 6, 7))
graph.addNode(new Node('wall', 7, 7))

// mark it as complete, which will calculate heuristics and add edges
graph.initialize()

// use it
graph.solve()

console.log('solution found:')

var solutionParent;
while (solutionParent = (solutionParent ? solutionParent.parent : graph.goalNode)) {
  console.log("coordinates are: (" + solutionParent.x + ", " + solutionParent.y + ")")
}

